### DISCUSSION TITLE
Magit rebase autosquash fixup does not seem to work correctly

### DISCUSSION THREAD

**Discussion by jmckitrick** [magit/magit#5232] [Support]
I'm running the latest emacs 31 and the latest Magit.

In a project, I have several commits to rebase, and one is a fixup and is marked 'fixup!'. It points to an existing commit made earlier in the day.

When viewing the log, I press 'r' then 'f', and nothing happens. I've tried this with the cursor on the parent, the fixup target, the fixup itself, and other random commits. The command being run according to the Messages buffer is something like 'git rebase -i --autosquash --keep-empty --autostash 8691788ab886d88f78d0dbb59d6f0841a1e6ed38' where that commit hash is the most recent one at the top of the log.

I was expecting to see the fixup commit disappear after being squashed into the target commit, or at least be dropped into an interactive rebase session.

Poking around the Magit code, I see a message that never appears while using the command: 'Type %p on a commit to squash into it and then rebase as necessary,' so this code is not reached for some reason.

What am I doing wrong?

**Comment by jmckitrick** [comment]
I think I found the answer... there's a different 'autosquash' option. Not sure what 'f' does but I think the other option might be what I want.

**Comment by kyleam** [comment]
> The command being run according to the Messages buffer is something like 'git rebase -i --autosquash --keep-empty --autostash 8691788ab886d88f78d0dbb59d6f0841a1e6ed38' where that commit hash is the most recent one at the top of the log.
>
> I was expecting to see the fixup commit disappear after being squashed into the target commit, [...]

That expectation is correct.  As you mention above, you see a rebase being executed with `--autosquash`.

With the information you provided, my guess is that commits between the base being chosen and HEAD doesn't include the target commit.

`magit-rebase-autosquash` invokes `magit-rebase-interactive-1` with a commit of `:merge-base`.  That triggers the following logic:

```elisp
    (if (eq commit :merge-base)
        (setq commit
              (and-let* ((upstream (magit-get-upstream-branch)))
                (magit-git-string "merge-base" upstream "HEAD")))
      ...)
```

You say the commit hash (i.e. base) is coming out to HEAD.  That would happen if your upstream branch either resolves to or contains HEAD, leading to HEAD as the base (a noop for squashing operation).

You could confirm that by instrumenting `magit-rebase-interactive-1`, calling `magit-rebase-autosquash`, and stepping through `magit-rebase-interactive-1`.
